npm_helpers.py

```diff
@@ -0,0 +1,103 @@
+    package_path = os.path.join(root_dir, 'package.json')
+    try:
+        with open(package_path, 'w') as f:
+            json.dump({
+                "name": "cloudpebble-ycmd-proxy",
+                "version": "1.0.0",
+                "description": "A temporary package.json",
+                "repository": None,
+                "license": "",
+                "dependencies": dependencies
+            }, f)
+        yield
+    finally:
+        os.unlink(package_path)
+
+
+def validate_dependencies(dependencies):
+    """ Check that none of the version strings in a dictionary of dependencies reference local paths. """
+    # CloudPebble performs identical checks for this, so hopefully it should never actually get triggered.
+    for version in dependencies.values():
+        if re.match(r'^file:|(\.*|~)/', version):
+            raise ValueError("Dependencies are not allowed to reference paths")
+
+
+def install_dependencies(dependencies, root_dir):
+    """ Install some npm dependencies into a directory and then extract the headers
+    :param dependencies: a dictionary of name->version
+    :param root_dir: the directory to install into
+    """
+    validate_dependencies(dependencies)
+    # Make a minimal package.json file
+    with temporary_package_json(root_dir, dependencies):
+        try:
+            # Install all the dependencies
+            # TODO: Should NPM itself have resource limits?
+            # TODO: Should we prune, or delete node_modules?
+            subprocess.check_output([settings.NPM_BINARY, "prune"], stderr=subprocess.STDOUT, cwd=root_dir)
+            if dependencies:
+                subprocess.check_output([settings.NPM_BINARY, "install", "--ignore-scripts"], stderr=subprocess.STDOUT, cwd=root_dir)
+        except subprocess.CalledProcessError:
+            # Setting the error message to e.output here would let the user see their lovely NPM error output.
+            raise NPMInstallError("One or more of your dependencies cannot be installed. Please check that the names and versions for all of your dependencies are valid.")
+
+
+def get_package_metadata(root_dir):
+    """ Given a directory with a node_modules directory full of pebble modules, get all of their messageKeys and resource type/name information
+    :param root_dir:
+    :return: (array of resources, array of (type, name) tuples
+    """
+    resources = []
+    messagekeys = []
+    try:
+        for package_path in glob.glob(os.path.join(root_dir, 'node_modules', '*', 'package.json')):
+            with open(package_path, 'r') as f:
+                data = json.load(f)
+                messagekeys.extend(data['pebble'].get('messageKeys', []))
+                resources.extend(Resource(r['type'], r['name']) for r in data['pebble'].get('resources', {}).get('media', []))
+        return resources, messagekeys
+    except Exception:
+        raise NPMInstallError("One or more of your dependencies is not a valid pebble library.")
```

Comment:
Ah, OK. I think removing the try/catch and adding `if 'pebble' not in data: continue` would be the right thing to do here then